### PR TITLE
Organizations from groups

### PR DESCRIPTION
**Pull request by jahow**
Add `groups` strategy alongside the default (`metadata`).

image

### CODE REVIEW

File: libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    map((groups) => {
+      return this.mapGroups(groups)
+    }),
+    shareReplay()
+  )
+  organisationsCount$ = this.organisations$.pipe(map((orgs) => orgs.length))
+
+  constructor(private groupService: GroupService) {}
+
+  equalsNormalizedStrings(
```
Comment (fgravin) [inline libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.ts]: duplicated & not used

File: libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    replaceSpecialChars = true
+  ): boolean {
+    if (!str1 || !str2) return false
+    return (
+      this.normalizeString(str1, replaceSpecialChars) ===
+      this.normalizeString(str2, replaceSpecialChars)
+    )
+  }
+
+  normalizeString(str: string, replaceSpecialChars = true): string {
```
Comment (fgravin) [inline libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.ts]: duplicated & not used

File: libs/feature/catalog/src/lib/feature-catalog.module.ts [hunk 3]
```diff
@@ -26,5 +31,29 @@ import { UiElementsModule } from '@geonetwork-ui/ui/elements'
...
+  providers: [
+    {
+      provide: OrganisationsServiceInterface,
+      useClass: OrganisationsFromMetadataService,
+    },
+  ],
 })
-export class FeatureCatalogModule {}
+export class FeatureCatalogModule {
+  static withOrganizationsStrategy(
```
Comment (fgravin) [inline libs/feature/catalog/src/lib/feature-catalog.module.ts]: maybe would need some documentation on how to change the strategy within your app ?

Comment (jahow) [reply libs/feature/catalog/src/lib/feature-catalog.module.ts]: Yes definitely! I will recreate this PR on upstream because nothing will be specific to geocat

File: libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      ],
+    },
+  },
+}
+
+class SearchApiServiceMock {
+  search = jest.fn(() => of(groupsAggregationMock))
+}
+
+const groupsApiMock = [
```
Comment (fgravin) [inline libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.spec.ts]: Don't we have fixtures for that ?
I think it's better to centralize in case of API changes.

File: libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  organisationsCount$ = this.organisations$.pipe(map((orgs) => orgs.length))
+
+  constructor(
+    private esService: ElasticsearchService,
+    private searchApiService: SearchApiService,
+    private groupService: GroupService,
+    private translateService: TranslateService
+  ) {}
+
+  private fetchGroupsAggregation() {
```
Comment (fgravin) [inline libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.ts]: Just a question, wasn't it the purpose of the `FieldService` ?

File: libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                size: 5000,
+              },
+            },
+          })
+        )
+      )
+      .pipe(map((response) => response.aggregations.groups.buckets))
+  }
+
+  private mapGroups(groupBuckets: any[], groups: GroupApiModel[]) {
```
Comment (fgravin) [inline libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.ts:67]: Adding the return type here helps for clarity and quickly understanding what does this function do.

File: libs/feature/search/src/lib/utils/service/fields.service.ts [hunk 7]
```diff
@@ -15,7 +20,10 @@ import {
 })
 export class FieldsService {
   private fields = {
-    publisher: new SimpleSearchField('OrgForResource', 'asc', this.injector),
+    publisher:
```
Comment (fgravin) [inline libs/feature/search/src/lib/utils/service/fields.service.ts]: I wonder if it wouldn't be better to have a `OrganizationSearchField` which would handle that, keeping the root service lighter. It's a bit intrusive here, the `if` statement and the injection.

**Review by fgravin — COMMENTED** [review]
Thanks for the good work.
I've made some minor comments.
I still have some concerns though
1. on home page loading, it queries twice the `groups`, might be an issue with the `GroupService.group$` observable indeed (my bad).
2. a click on an organization in the organization tab does not filter on the org
3. a click on the contact in the metadata detailed page does not filter on the org
4. I find the last commit quite intrusive. I think we should keep the mapping synchronous, and handle the async hydrating objects ahead of time, wdyt ?

Also, it's just a thought, but I wonder if the org service shouldn't centralize the whole things. The PR affects many places in the code while I think it's better if the strategy is implemented in one place only. wdyt ?

Sorry for this review :smirk: 

Also, in terms of looking, we should try to not stretch the logos. The borders are also truncated .. But not really part of the work here

File: libs/feature/search/src/lib/utils/service/fields.service.ts (hunk 7 above)
Comment (jahow) [reply libs/feature/search/src/lib/utils/service/fields.service.ts]: I don't know, both fields have different implementations so the `OrganizationSearchField` would contain a lot of if's...

File: libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.ts (hunk 5 above)
Comment (jahow) [reply libs/feature/catalog/src/lib/organisations/service/organisations-from-groups.service.ts]: Not really, the `FieldService` is there to offer an abstraction on the search index and allow filtering and aggregating with agnostic field names.

The `OrganisationsFromGroupsService` also aggregates by group owner but this is an implementation detail, its only job is to output a list of organisations. I mean technically it does something very similar, but here we're not really doing anything related to _search_ in the sense of _searching records_. Another implementation could just query a `/api/groups` endpoint for example and that's it.

And speaking of which, the more I think about it the more I feel like the `feature-catalog` should not depend on the `feature-search` module at all; if anything, the dependency should be the other way around.

**Comment by jahow** [comment]
> 1. on home page loading, it queries twice the `groups`, might be an issue with the `GroupService.group$` observable indeed (my bad).

I will look into it. I'm not really sure about the added value of this service though, compared to the `GroupsApiService`?

>     2. a click on an organization in the organization tab does not filter on the org
>
>     3. a click on the contact in the metadata detailed page does not filter on the org

Yep, I still have to fix that one.
 
>     4. I find the last commit quite intrusive. I think we should keep the mapping synchronous, and handle the async hydrating objects ahead of time, wdyt ?

I'm not sure, what would be the reason for keeping the mapping synchronous? This feels like a constraint that would force us to implement several steps of mappings (or "hydration"), where we could just have one.

Also later on this will be "hidden" in the RecordsRepository implementation, which will obviously provide records in an async way.

> Also, it's just a thought, but I wonder if the org service shouldn't centralize the whole things. The PR affects many places in the code while I think it's better if the strategy is implemented in one place only. wdyt ?

Yes I really agree on that, I wish I could use the `feature-catalog` lib in `feature-search`... This might take me too far in terms of refactoring though.

**Comment by jahow** [comment]
(thanks for the review :slightly_smiling_face: )

**Review by fgravin — COMMENTED** [review]
What is the status of this PR, do you need another review or is it ok ?

**Comment by jahow** [comment]
> What is the status of this PR, do you need another review or is it ok ?

Working on making the orgs service used everywhere which is a lot of refactoring; I will recreate this PR on upstream

**Comment by fgravin** [comment]
> > What is the status of this PR, do you need another review or is it ok ?
> 
> Working on making the orgs service used everywhere which is a lot of refactoring; I will recreate this PR on upstream

Ok, the catalog lib could be shared by anyone doesn't it ?
We will need organisation in the search, in the apps, in the editor, in the admin etc...

**Comment by jahow** [comment]
I've updated the branch with my WIP changes; missing some tests and there are still some bugs in there.

Basically the org service (and its different implementations) provides a list of orgs and will also read the organizations from an ES source document and generate ES filters from orgs. What's impactful is that many things became asynchronous in the process...

File: apps/datahub/src/app/home/organisations-page/organisations-page.component.ts [hunk 8]
```diff
@@ -1,9 +1,23 @@
...
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class OrganisationsPageComponent {}
+export class OrganisationsPageComponent {
+  constructor(
+    private searchService: SearchService,
+    private orgsService: OrganisationsServiceInterface
+  ) {}
+
+  searchByOrganisation(organisation: Organisation) {
```
Comment (fgravin) [inline apps/datahub/src/app/home/organisations-page/organisations-page.component.ts:18]: Note that this will conflict the 'right-click' work done by @tkohr and @f-necas

File: libs/feature/catalog/src/lib/organisations/organisations.component.html [hunk 9]
```diff
@@ -12,7 +12,7 @@
   >
     <gn-ui-organisation-preview
       [organisation]="organisation"
-      (clickedOrganisation)="searchByOrganisation($event)"
+      (clickedOrganisation)="orgSelect.next($event)"
```
Comment (fgravin) [inline libs/feature/catalog/src/lib/organisations/organisations.component.html]: `emit` is clearer

File: apps/datahub/src/app/home/organisations-page/organisations-page.component.html [hunk 10]
```diff
@@ -1,3 +1,3 @@
 <div class="container-lg mx-auto mt-8">
-  <gn-ui-organisations></gn-ui-organisations>
+  <gn-ui-organisations (orgSelect)="(searchByOrg)"></gn-ui-organisations>
```
Comment (fgravin) [inline apps/datahub/src/app/home/organisations-page/organisations-page.component.html]: `searchByOrganisation` instead ?

**Review by fgravin — COMMENTED** [review]
:scream: 

Hard to follow the whole thing. But I think I got the point. Maybe, it would be great to add some words explaining the contact hydration process.

How the whole thing - metadata contact - groups - index - publisher - organisation tab - are all connected together :smile: 

Very good though :+1:

**Comment by jahow** [comment]
> scream
> 
> Hard to follow the whole thing. But I think I got the point. Maybe, it would be great to add some words explaining the contact hydration process.
> 
> How the whole thing - metadata contact - groups - index - publisher - organisation tab - are all connected together smile
> 
> Very good though +1

Yes I mean an explanation of all this would definitely have its place in the (soon to be) technical documentation of the project. As with many things in gn-ui some pieces of the architecture have been thought to play a certain role in the grand scheme of things, and just looking at the code might not be enough to get the full picture.

But in any case this refactoring should lead to less code duplication and a concept of organizations that is more consistent throughout the apps.

**Comment by jahow** [comment]
Created https://github.com/geonetwork/geonetwork-ui/pull/508 to integrate these changes upstream, so I'm closing this.
